[Accessibility] Make screen readears announce stock quantity when selecting a product variation

Manussakis opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #52141  .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set up a variable product with size variations.
![image](https://github.com/user-attachments/assets/0f26a803-5cb3-466b-b518-b180f3aceca0)
2. For each one of the variations add a different stock quantity. 
![Screenshot 2025-01-28 at 10 42 01](https://github.com/user-attachments/assets/70624da7-2751-43d3-8d03-0c3e20f47d55)
3. Turn on the screen reader.
3. Go to the product page on the frontend and select each one of the sizes.
4. Verify the screen reader announces the stock quantity.
![image](https://github.com/user-attachments/assets/080951dd-b904-483d-aa69-1cda5103733b)
5. Test on different browsers.
6. Test on Storefront and WP default themes.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Manussakis committed:
Make screen readears announce stock quantity when selecting a product variation

Manussakis committed:
Add changelog file

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/add-to-cart-variation.js:
```diff
@@ -304,7 +304,11 @@
 		$template_html = $template_html.replace( '/*]]>*/', '' );
 
-		form.$singleVariation.html( $template_html );
+		// Add a delay before updating the live region to ensure screen readers pick up the content changes.
+		setTimeout( function() {
+			form.$singleVariation.html( $template_html );
+		}, 500);
```
This is especially needed when using Firefox + VoiceOver.

Manussakis committed:
Merge branch 'trunk' into fix/52141

Manussakis committed:
Merge branch 'trunk' into fix/52141

Manussakis committed:
Group template population and reveal

Manussakis committed:
Add waitForTimeout once variation is selected in tests

Manussakis committed:
Merge branch 'fix/52141' of github.com:Manussakis/woocommerce into fix/52141

Manussakis committed:
Merge branch 'trunk' into fix/52141

Manussakis commented on plugins/woocommerce/tests/e2e-pw/tests/product/product-variable.spec.js:
```diff
@@ -317,6 +317,8 @@ test.describe(
 
 			await page.locator( '#colour' ).selectOption( 'Red' );
 
+			await page.waitForTimeout( 300 );
+
```
Some tests find the `.woocommerce-variation-price` element to get its text content once a product variation is selected. However, they started to fail after implementing the delay of 300 ms before updating the DOM.

The solution was to add `await page.waitForTimeout( 300 );` right after every time the product variation is selected.

Manussakis commented:
@dkotter This PR is ready for code review.
One e2e test failed after the last build, but it seems flaky since the same test passed before.

dkotter reviewed (approved):
@ankitguptaindia @qasumitbagthariya This is ready for QA

qasumitbagthariya commented:
@Manussakis ⚠️

I have a question regarding screen reader behavior. Based on my testing, I created a variable product with one attribute and two attributes (e.g., color, size, and color). In this case, the screen reader announces the stock quantity along with the price. Is this expected behavior, or should it only announce the stock quantity?

https://github.com/user-attachments/assets/461ad79d-2d82-4576-8725-4e2ecca2af5c

Manussakis commented:
@qasumitbagthariya 

> I have a question regarding screen reader behavior. Based on my testing, I created a variable product with one attribute and two attributes (e.g., color, size, and color). In this case, the screen reader announces the stock quantity along with the price. Is this expected behavior, or should it only announce the stock quantity?

This is expected. Any update related to description, price, and availability that changes once the product variation is selected should be announced since they are inside the same live region.

![image](https://github.com/user-attachments/assets/226a8ba8-101c-4d7f-9bc9-5158ab556ac3)

qasumitbagthariya commented:
### QA/Test Report ✅ 
-------------

Thanks for the comment. I'm approving this PR and ready for the next step.

@Manussakis please remove [WIP] from the title and take this PR out of the draft.

vikrampm1 commented:
@ObliviousHarmony this PR is ready for your review!

jeffpaul commented:
@ObliviousHarmony reminder ping for your code review on this accessibility PR, thanks!

ObliviousHarmony committed:
Merge branch 'trunk' into fix/52141

ObliviousHarmony reviewed (changes requested):
Thanks @Manussakis! I'm sorry for the wait, I just got back from sabbatical and have missed some GitHub notifications since then! I've taken a look at the PR and I've got some thoughts.

- [There are some failing E2E tests now](https://github.com/woocommerce/woocommerce/actions/runs/13251943078/job/36992418503?pr=54923#step:9:325). I re-tried it multiple times to see if it was flaky and updated the branch to confirm that it wasn't an issue that was fixed on `trunk`.
- In theory, could introducing a 300ms delay between the other jQuery event triggers and the `show_variation` trigger be a breaking change? I'm not sure if it's something to even be concerned about but it's food for thought.

Manussakis committed:
Merge branch 'trunk' into fix/52141

Manussakis committed:
Add delay to tests that select product variations

Manussakis commented:
Hey @ObliviousHarmony 👋🏽 

> There are some failing E2E tests now. I re-tried it multiple times to see if it was flaky and updated the branch to confirm that it wasn't an issue that was fixed on trunk.

Thanks for catching that. I have fixed it.

---

> In theory, could introducing a 300ms delay between the other jQuery event triggers and the show_variation trigger be a breaking change? I'm not sure if it's something to even be concerned about but it's food for thought.

That's a fair point 🤔 .

It is hard to guarantee, but in theory, there shouldn't be a breaking change because:
- The `show_variation` is the last event triggered inside the `onFoundVariation` function, so there is no other event that could depend on this one to be called first.
- The [`onShow`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/frontend/add-to-cart-variation.js#L101) callback, that fires once the `show_variation` event goes off, doesn't trigger any other event that could be impacted by being delayed.

This is the theory, however, it's hard to predict what plugins and themes are implementing when they hook into the `show_variation` event.

Triggering the `show_variation` without waiting for the 300ms delay doesn't seem to be a good idea too, because the variation information is not in the DOM yet.

vikrampm1 commented:
@ObliviousHarmony this is back for your re-review.

ObliviousHarmony committed:
Merge branch 'trunk' into fix/52141

ObliviousHarmony reviewed (approved):
Sounds good to me @Manussakis :)
